### PR TITLE
 Added user registration and account verification with email

### PR DESCRIPTION
**Pull request by Rezdoom**

### CODE REVIEW

File: server/Domain/Services/ICryptoService.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export interface ICryptoService {
+  encrtyptAES(data: string): string;
+  decryptAES(data: string): string;
+  encrtyptSHA256(data: string): any;
```
Comment (CeQadrat) [inline server/Domain/Services/ICryptoService.ts]: encrypSHA256 return string, don't any

File: server/Data/Repositories/IRepository.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@ export interface IRepository<T> {
 
   findAll(filter: any): Promise<T[]>;
 
-  find(filter: any): Promise<T[]>;
+  find(filter: any): Promise<T>;
```
Comment (CeQadrat) [inline server/Data/Repositories/IRepository.ts]: Rename find to findOne

File: server/Domain/Services/Impl/UserService.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  ) {
+    this._userRepository = objectiveRepository;
+    this._cryptoService = cryptoService;
+    this._mailerService = mailerService;
+  }
+
+  public async registerUser(data: any) {
+    const userData = {
+      ...data,
+      Password: this._cryptoService.encrtyptSHA256(data.Password),
```
Comment (paveldz) [inline server/Domain/Services/Impl/UserService.ts]: User passwords must be hashed, not encrypted. Also please use salt for hashing passwords

File: server/API/controllers/UserController.ts [hunk 4]
```diff
@@ -5,13 +5,38 @@ import {
...
+  }
+
+  @httpGet('/verifyEmail/:hash')
+  private async verifyEmail(@response() res: express.Response, @requestParam('hash') hash: string): Promise<void> {
+    if (await this._userService.verifyEmail(decodeURIComponent(hash))) {
+      // redirect to front page
+      res.redirect('https://vk.com/');
+    }
+    // link is not valid
+    res.redirect('https://ok.ru/');
```
[the remark this answers is not in the document]
Comment (Rezdoom) [reply server/API/controllers/UserController.ts]: when the front page is created, then there will be a redirect to it

File: server/Domain/Services/Impl/UserService.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    @inject('MailerService') mailerService: IMailerService,
+  ) {
+    this._userRepository = objectiveRepository;
+    this._cryptoService = cryptoService;
+    this._mailerService = mailerService;
+  }
+
+  public async registerUser(data: any) {
+    const userData = {
+      ...data,
```
[the remark this answers is not in the document]
Comment (Rezdoom) [reply server/Domain/Services/Impl/UserService.ts]: no, only hashed password

**Pull request merged**